pandas_gbq/gbq.py

```diff
@@ -577,141 +500,97 @@ def run_query(self, query, **kwargs):
-
-            try:
-                query_reply = job_collection.getQueryResults(
-                    projectId=job_reference['projectId'],
-                    jobId=job_id,
-                    pageToken=page_token).execute()
-            except HttpError as ex:
-                self.process_http_error(ex)
-
-        if current_row < total_rows:
-            raise InvalidPageToken()
+        try:
+            rows_iter = query_reply.result()
+        except self.http_error as ex:
+            self.process_http_error(ex)
+        result_rows = list(rows_iter)
+        total_rows = rows_iter.total_rows
+        schema = {
+            'fields': [
+                field.to_api_repr()
+                for field in rows_iter.schema],
+        }
 
         # print basic query stats
         self._print('Got {} rows.\n'.format(total_rows))
 
-        return schema, result_pages
+        return schema, result_rows
 
     def load_data(self, dataframe, dataset_id, table_id, chunksize):
-        try:
-            from googleapiclient.errors import HttpError
-        except ImportError:
-            from apiclient.errors import HttpError
+        from google.cloud.bigquery import LoadJobConfig
+        from six import StringIO
 
-        job_id = uuid.uuid4().hex
+        destination_table = self.client.dataset(dataset_id).table(table_id)
+        job_config = LoadJobConfig()
+        job_config.write_disposition = 'WRITE_APPEND'
+        job_config.source_format = 'NEWLINE_DELIMITED_JSON'
         rows = []
         remaining_rows = len(dataframe)
 
         total_rows = remaining_rows
         self._print("\n\n")
 
         for index, row in dataframe.reset_index(drop=True).iterrows():
-            row_dict = dict()
-            row_dict['json'] = json.loads(row.to_json(force_ascii=False,
-                                                      date_unit='s',
-                                                      date_format='iso'))
-            row_dict['insertId'] = job_id + str(index)
-            rows.append(row_dict)
+            row_json = row.to_json(
+                force_ascii=False, date_unit='s', date_format='iso')
+            rows.append(row_json)
             remaining_rows -= 1
 
             if (len(rows) % chunksize == 0) or (remaining_rows == 0):
-                self._print("\rStreaming Insert is {0}% Complete".format(
+                self._print("\rLoad is {0}% Complete".format(
                     ((total_rows - remaining_rows) * 100) / total_rows))
 
-                body = {'rows': rows}
+                body = StringIO('{}\n'.format('\n'.join(rows)))
 
                 try:
-                    response = self.service.tabledata().insertAll(
-                        projectId=self.project_id,
-                        datasetId=dataset_id,
-                        tableId=table_id,
-                        body=body).execute()
-                except HttpError as ex:
+                    self.client.load_table_from_file(
```

Comment:
This is so much better than the existing method

Reply:
Yeah, it's technically a change in behavior (kicks off a load job instead of using the streaming API), but I think the change is small enough to be worth it. Load jobs should be much more reliable for the use case of this library.